### ISSUE TITLE
Possible payments state machine improvements

### ISSUE THREAD

**Issue by pierre** [killbill/killbill#330]
* Ability to export per-plugin state machines (could also be in an external system, do we need a special plugin type?)
* Ability to version state machines (e.g. PSP starts providing refund capability for future payments)
* Ability to override state transitions (e.g. PSP sends REFUND_SUCCESS but when the acquirer actually processes it, it fails, and the PSP sends a REFUND_FAILURE for that same refund)

**Comment by sbrossie** [comment]
+1 for Ability to override state transitions: I think we also need some kind of APIs to fix state associated with bugs; maybe this is applicable for that case as well.

**Comment by pierre** [comment]
> * Ability to override state transitions (e.g. PSP sends REFUND_SUCCESS but when the acquirer actually processes it, it fails, and the PSP sends a REFUND_FAILURE for that same refund)

Fixed by https://github.com/killbill/killbill/issues/332.

**Comment by pierre** [comment]
🚿 💭 

It seems that this would be solved by making the state machine XML configurable via a per-tenant, per-plugin, tenant_kvs entry (similar mechanism than the plugin config). Versioning and caching would pretty much come for free.

Only new JAX-RS and internal APIs would be required, so it could even land in 0.16.

**Comment by sbrossie** [comment]
I understand the per-plugin state machine use case, but do we really need per tenant? I guess maybe there are use cases where a gateway could be configured differently for different tenants and therefore leading to the per-tenant, per-plugin scenario?

**Comment by pierre** [comment]
> I understand the per-plugin state machine use case, but do we really need per tenant?

I think it needs to be similar to the plugin configuration (which is per-tenant today): different merchant accounts would be associated to potentially different processors configured in the gateway (hence different behaviors and different state machines).

**Comment by sbrossie** [comment]
Commenting on this ticket for the state machine improvement and also with regard to open ticket #504 and #514:

In order to 'fully" address those related issues, we need the following:

1. Per per-plugin (and also maybe per-tenant) state machines
2. Some transitions should only be allowed with respect to previous *same* transaction (PENDING use case) [maybe a **'sticky'** attribute in the state machine] to avoid for instance that [kind of logic](https://github.com/killbill/killbill/blob/killbill-0.16.4/payment/src/main/java/org/killbill/billing/payment/core/sm/payments/PaymentLeavingStateCallback.java#L91).
3. Terminal state for a transaction [maybe a **'terminal'** attribute in the state machine]. For instance to avoid situations like `AUTH` + `VOID` succeeds but then  a `CAPTURE`  would succeed because the transition `VOID_SUCCESS` -> `CAPTURE_INIT` is allowed
4. Linking transactions. Today such a use case is possible `AUTH`, `CAPTURE`, `VOID`, `VOID`, but we lose track of which `VOID` is linked to which initial operation. This make things more difficult to understand what's next.
5. We should review our state machine description. For instance i am not understanding why [this is defined](https://github.com/killbill/killbill/blob/killbill-0.16.4/payment/src/main/resources/org/killbill/billing/payment/PaymentStates.xml#L468) as a `linkStateMachine` (should be `transition`)

**Comment by pierre** [comment]
> Terminal state for a transaction [maybe a 'terminal' attribute in the state machine].

Terminal attribute on a transition is a great idea! 👍 

> Linking transactions. Today such a use case is possible `AUTH`, `CAPTURE`, `VOID`, `VOID`, but we lose track of which `VOID` is linked to which initial operation. This make things more difficult to understand what's next.

I don't think we can actually track them with our payment API. `AUTH`, `CAPTURE`, `VOID`, `CAPTURE`, `VOID` is ambiguous: it either means both captures were cancelled and the amount is still authorized or the second capture went through and we reversed the originally auth (partial capture use-case). I bet that sometimes the plugin won't even know exactly what's happening because the gateway API is ambiguous too. I did look at stack-based finite-state machines to solve that problem (what we're doing implicitly today inside Kill Bill already) and this is why I pivoted to Mealy machines.

**Comment by sbrossie** [comment]
> `AUTH`, `CAPTURE`, `VOID`, `CAPTURE`, `VOID` is ambiguous:

Is it really (assuming we know result of the operation when it returns, which is mostly the case)? 

1. You start with the `AUTH`
2. Follow with the `CAPTURE` (Let's assume this was successful)
3. Then issue the `VOID`. At this point the `VOID` would either succeed to reverse the `CAPTURE` (that was not sent for settlement yet), or fail.
4. The following `CAPTURE` then only makes sense if that previous `VOID` succeed
5. Finally the last `VOID` would either work against the `CAPTURE` if it succeeded and again assuming that was not sent for settlement yet or against the original `AUTH`.

**Comment by pierre** [comment]
> Is it really (assuming we know result of the operation when it returns, which is mostly the case)?

Given our `PaymentPluginApi` signature, I believe so:

```
public PaymentTransactionInfoPlugin voidPayment(UUID kbAccountId, UUID kbPaymentId, UUID kbTransactionId, UUID kbPaymentMethodId, Iterable<PluginProperty> properties, CallContext context) throws PaymentPluginApiException;
```

Eventually, the plugin determines what to do with the void request, see https://github.com/killbill/killbill-plugin-framework-ruby/blob/master/lib/killbill/helpers/active_merchant/active_record/models/transaction.rb#L36-L54.

Note that we only have one usecase today (CyberSource), maybe it would be interesting to see if other gateways support that.

> The following `CAPTURE` then only makes sense if that previous `VOID` succeed

Not necessarily: the initial capture could be partial (`AUTH $10 SUCCESS`, `CAPTURE $5 SUCCESS`, `VOID (of previous capture) FAILED`, `CAPTURE $5 SUCCESS`).

**Comment by pierre** [comment]
To summarize this issue:

* https://github.com/killbill/killbill/issues/332 was created and fixed to allow administrators to override state transitions.
* I will implement per-plugin per-tenant state machines.
* Instead of introducing a terminal attribute on transitions, I will revert changes in the default state machine to allow void on capture (https://github.com/killbill/killbill/issues/504). We only have one use-case today (CyberSource - although I believe Litle supports it as well), and it complicates the generic logic. Users who use CyberSource and use the VOID on capture functionality can simply update their own state machine config to allow the transition.

Won't fix at this point:

* True versioning of state machines. The original use-case was PSP starting to provide refund capabilities for payments starting at time T1. With a versioning scheme similar to the catalog versioning, we could have prevented refund on a payment effective at T0 < T1. With the proposed implementation, the state machine configuration can be updated to reflect that refunds are now allowed but it won't enforce that they are disallowed for payments before T1. A work-around is to implement this logic in the plugin (one way would be to interpret the associated error from the PSP and return a `CANCELED` state).
* Logic outside of state machine ([UNKNOWN transactions handling](https://github.com/killbill/killbill/blob/3559fcc4309701759032c4059059b0127142faac/payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java#L381) and [sanity on preventing multiple inital PENDING transactions](https://github.com/killbill/killbill/blob/3559fcc4309701759032c4059059b0127142faac/payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java#L416-L422)).

**Comment by pierre** [comment]
Implemented in branches work-for-release-0.16.7 of the various repos. I've also updated our docs with the default state machine: http://docs.killbill.io/0.16/userguide_payment.html#_payment_states.
